### PR TITLE
Screeps decoration update new features and deprecations

### PR DESCRIPTION
**Pull request by shanemadden**
New features for Screeps 4.2 currently on PTR, not ready for merge but able to be used for anyone testing on the PTR.

Currently just the 3 new intershard resources and the new function and constant for pixel generation - I expect there will be more changes to come.

### CODE REVIEW

**Comment by shanemadden** [comment]
Updated and works on the released MMO version, but the devs won't be adding stub functions for the cpu unlock and pixel interfaces, so we'll want those to fail in safer ways than currently implemented.

File: javascript/utils.js [hunk 1]
```diff
@@ -207,6 +207,9 @@ function __resource_type_num_to_str(num) {
         case 83: return RESOURCE_EMANATION;
         case 84: return RESOURCE_ESSENCE;
         case 1001: return SUBSCRIPTION_TOKEN;
+        case 1002: return CPU_UNLOCK;
```
Comment (ASalvail) [inline javascript/utils.js:210]: Is it just me or are all those constants gone from the screeps doc? We could just now encode this in an enum instead.

Not in this PR, but still.

File: src/game/map.rs [hunk 2]
```diff
@@ -164,13 +163,12 @@ pub fn find_route_with_callback(
 
     // Type erased and boxed callback: no longer a type specific to the closure
     // passed in, now unified as &Fn
-    let callback_type_erased: &mut (dyn FnMut(RoomName, RoomName) -> f64) =
-        &mut callback_boxed;
+    let callback_type_erased: &mut (dyn FnMut(RoomName, RoomName) -> f64) = &mut callback_boxed;
```
Comment (ASalvail) [inline src/game/map.rs:166]: Do you have a special `rustfmt` config? Line length should be pretty standard; I don't get why it's jumping around.

File: src/game/market.rs [hunk 3]
```diff
@@ -1,303 +1,304 @@
...
+/// Provide a resource type to get history for using `Some(ResourceType)`, or
+/// get data for all resources by passing `None`
+pub fn get_history(resource: Option<MarketResourceType>) -> Vec<OrderHistoryRecord> {
+    match resource {
+        Some(resource_type) => {
+            match resource_type {
+                // workaround: Game.market.getHistory returns `{}` instead of `[]` when querying a resource type
+                // that has no history records
+                // Verify records are present otherwise return an empty array to prevent panics
+                // Workaround no longer needed if this PR gets merged: https://github.com/screeps/engine/pull/131
```
Comment (ASalvail) [inline src/game/market.rs:276]: Can we track this in an issue instead of in the code?

**Review by ASalvail — APPROVED** [review]

File: javascript/utils.js (hunk 1 above)
Comment (shanemadden) [reply javascript/utils.js:210]: Hrmm not sure if I'm following - should still be in their docs? https://github.com/screeps/common/blob/master/lib/constants.js#L379

File: src/game/map.rs (hunk 2 above)
Comment (shanemadden) [reply src/game/map.rs:166]: Nothing custom, I'm just gonna keep blaming nightly builds ;)

File: javascript/utils.js (hunk 1 above)
Comment (ASalvail) [reply javascript/utils.js:210]: You're right, I checked their history and I imagined things. I thought the integers were related to similar constants in the js implementation. It seems we only use them to make the translation easier.

**Pull request merged**